hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.monitors;
+
+import com.hazelcast.instance.GroupProperties;
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.ConnectionManager;
+import com.hazelcast.nio.OutboundFrame;
+import com.hazelcast.nio.Packet;
+import com.hazelcast.nio.tcp.TcpIpConnection;
+import com.hazelcast.nio.tcp.TcpIpConnectionManager;
+import com.hazelcast.nio.tcp.nonblocking.NonBlockingSocketWriter;
+import com.hazelcast.nio.tcp.spinning.SpinningSocketWriter;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.operationservice.impl.operations.Backup;
+import com.hazelcast.util.MutableInteger;
+
+import java.text.NumberFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Random;
+import java.util.Set;
+
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_PERIOD_SECONDS;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_SAMPLES;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_THRESHOLD;
+import static java.lang.Math.min;
+import static java.util.Collections.emptySet;
+
+/**
+ * The OverloadedConnectionsPlugin checks all the connections and samples the content of the packet
+ * queues if the size is above a certain threshold. This is very useful to figure out when huge
+ * amount of memory is consumed due to pending packets.
+ *
+ * Currently the sampling has a lot of overhead since the value needs to be deserialized. That is why this
+ * plugin is disabled by default.
+ */
+public class OverloadedConnectionsPlugin extends PerformanceMonitorPlugin {
+    private static final Queue<OutboundFrame> EMPTY_QUEUE = new LinkedList<OutboundFrame>();
+
+    private final SerializationService serializationService;
+    private final ILogger logger;
+    private final Map<String, MutableInteger> occurrenceMap = new HashMap<String, MutableInteger>();
```

Comment:
a small suggestion: `ItemCounter` might be a more suitable abstraction for this. It miss iterations, but it would be easier to add.